### ISSUE TITLE
transmuting Vec might lead to undefined behavior

### ISSUE THREAD

**Issue by TheEdda** [Smithay/drm-rs#117]
Hello, I've come across these transmutes when browsing the code:

https://github.com/Smithay/drm-rs/blob/e5f5b14d58c0d598cbce4019c706b2b1e0cf7d7e/src/control/mod.rs#L170

https://github.com/Smithay/drm-rs/blob/e5f5b14d58c0d598cbce4019c706b2b1e0cf7d7e/src/control/mod.rs#L539

They are transmuting `Vec<drm_mode_modeinfo>` into `Vec<Mode>`. However, `Vec` has an unspecified layout and this can lead to UB. This is even mentioned in the documentation https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives (scroll to "Turning a `Vec<&T>` into a `Vec<Option<&T>>`.")

The documentation also offers an alternative:
```rs
let after = unsafe {
    let mut before = std::mem::ManuallyDrop::new(before);
    Vec::from_raw_parts(before.as_mut_ptr() as *mut _, before.len(), before.capacity())
};
```

Secondly, I have also notices these two `Handle` types aren't `repr(transparent)` as they should be. Other `Handle` types are, so I assume this is just an oversight. 

https://github.com/Smithay/drm-rs/blob/e5f5b14d58c0d598cbce4019c706b2b1e0cf7d7e/src/control/encoder.rs#L11

https://github.com/Smithay/drm-rs/blob/e5f5b14d58c0d598cbce4019c706b2b1e0cf7d7e/src/control/crtc.rs#L20

I can open a PR to fix these, should be very simple.

**Comment by Drakulix** [comment]
Both seem like good suggestions. Feel free to open a PR, I would gladly accept it.

**Comment by TheEdda** [comment]
I have opened a PR with these changed. Furthermore, I've been messing with my fork and decided to reduce the number of transmutes overall, just to feel a little bit better. Most of the time they can be avoided altogether, although with two caveats:
* It uses `array::map` which was added in Rust 1.55 - I don't know if you have any minimum rust version policy
* It _might_ have an effect on performance, depending on how the compiler decides to optimize it, but hopefully those paths aren't that hot anyway (plus compared to libdrm, which mallocs all over the place, I think this is still good).

The diff is here (it includes the PR I just opened as well) https://github.com/Smithay/drm-rs/compare/develop...TheEdward162:feature/reduce-transmutes

Would you be interested in merging this as well? If not I'll keep it in my fork.

**Comment by Drakulix** [comment]
We currently don't have a minimum version policy, but given that the crate currently is very conservative in it's use of newer APIs, I would bet some projects dependent on drm-rs might have.

I think a lot of your improvements should still apply without `array::map`, no?
For merging a PR with `array::map` we should probably have a discussion on adding a MSRV to the crate first and likely tag a new major-release just to be safe.

**Comment by TheEdda** [comment]
Many of them would, yes. Also, in this case, I think `array::map` could be emulated by a macro in most places.

**Comment by TheEdda** [comment]
I think the `bytemuck` crate would work with arrays and slices as well. Their MSRV is 1.34.

**Comment by Drakulix** [comment]
> I think the `bytemuck` crate would work with arrays and slices as well. Their MSRV is 1.34.

1.34 sounds good enough, I think that is definitely a reasonable target. 👍

**Comment by TheEdda** [comment]
I have investigated further and here's what I found:

The alloc feature on bytemuck requires Rust 1.36, so that's two versions higher than the 1.34 they have without that feature.


Unfortunately the [`cast_vec`](https://docs.rs/bytemuck/latest/bytemuck/allocation/fn.cast_vec.html) from bytemuck requires the types to implement some traits (`NoUninit` and `AnyBitPattern` exactly, but ideally just `Pod`). There is only one type that needs Vec casting - the `drm_ffi::drm_mode_modeinfo` struct and its transparent wrappper `Mode`. I don't think messing with `drm_ffi` and adding bytemuck derives is the best idea here, so I have [opened an issue](https://github.com/Lokathor/bytemuck/issues/110) and a PR to allow this functionality for repr(transparent) types.

Otherwise, there would need to be some bytemuck derives directly in the `drm_ffi` crate - this could be achieved by using the [`ParserCallbacks::add_derives`](https://docs.rs/bindgen/latest/bindgen/callbacks/trait.ParseCallbacks.html#method.add_derives) method and adding a bytemuck derive directly onto the struct as needed.

About the array mapping - to avoid the 1.55 MSRV this would need to be switched back to using some kind of transmute. However, this causes problems with the handle types because bytemuck and orphan rules don't allow us to implement `Pod` for `Option<Handle>` so that won't let us cast the arrays of u32 into arrays of optional handles. For that I have also [opened an issue](https://github.com/Lokathor/bytemuck/issues/112) but I'm not sure if that proposition is the best solution.
